operator/src/main/java/org/bf2/operator/managers/ResourceManager.java

```diff
@@ -0,0 +1,339 @@
+            message = String.format("The units for %s/%s was changed", managedKafka.getMetadata().getNamespace(),
+                    managedKafka.getMetadata().getName());
+        }
+
+        if (message != null) {
+            log.error(message);
+            return Optional.of(new OperandReadiness(Status.False, Reason.Error, message));
+        }
+
+        return Optional.empty();
+    }
+
+    public Optional<OperandReadiness> validateResources(ManagedKafka managedKafka) {
+        Optional<ManagedKafkaCondition> optReady = Optional.ofNullable(managedKafka.getStatus())
+                .flatMap(s -> ConditionUtils.findManagedKafkaCondition(s.getConditions(),
+                        ManagedKafkaCondition.Type.Ready));
+        if (optReady.filter(c -> ManagedKafkaCondition.Reason.Rejected.name().equals(c.getReason())).isPresent()) {
+            // rejected for a normal deployment is terminal
+            return Optional.of(new OperandReadiness(Status.False, Reason.Rejected, optReady.get().getMessage()));
+        }
+        ManagedKafkaAgent localAgent = informerManager.getLocalAgent();
+        // eventually we may want to require the local agent exists and is configured for max nodes
+        // but for now we're lax
+        if (localAgent == null) {
+            return Optional.empty();
+        }
+        String profile = KafkaInstanceConfigurations.getInstanceType(managedKafka);
+        ConfigMap resourceConfigMap = getOrCreateResourceConfigMap(localAgent);
+        if (resourceConfigMap == null) {
+            return Optional.empty();
+        }
+        String claimedResources = resourceConfigMap.getData().get(Cache.metaNamespaceKeyFunc(managedKafka));
+        if (claimedResources == null) {
+            return claimResources(managedKafka, profile, localAgent);
+        }
+        return validateResources(managedKafka, profile, claimedResources);
+    }
+
+    private ConfigMap getCachedResourceConfigMap() {
+        return informerManager.getLocalConfigMap(client.getNamespace(), ResourceManager.FLEETSHARD_RESOURCES);
+    }
+
+    public Map<String, ProfileCapacity> buildCapacity(ManagedKafkaAgent resource, ManagedKafkaAgentStatus status) {
+        Map<String, ProfileCapacity> capacity = new LinkedHashMap<>();
+
+        Map<String, Profile> profiles = resource.getSpec().getCapacity();
+
+        Optional<ConfigMap> resourceConfigMap = Optional.ofNullable(getCachedResourceConfigMap());
+
+        for (String key : profiles.keySet()) {
+            getProfileMaxUnits(profiles, key).ifPresent(max -> {
+                capacity.put(key,
+                        new ProfileCapacityBuilder().withMaxUnits(max)
+                                .withRemainingUnits(resourceConfigMap.map(c -> c.getData().get(key))
+                                        .map(Integer::valueOf)
+                                        .map(i -> max - i)
+                                        .orElse(null))
```

Comment:
My thinking here wasn't quite right - since the cached configmap is being used, it could be missing, in which case null is appropriate.  But that doesn't make sense if the map exists and just doesn't have an entry for that profile type.

It's probably better to use getOrCreateResourceConfigMap so that it's known that the configmap can't be null.  In that case the else should clearly be max.